syntax = "proto3";

package ecliptix.proto.protocol;
option csharp_namespace = "Ecliptix.Protobuf.Protocol";

import "google/protobuf/timestamp.proto";

message SecureEnvelope {
  uint32 version = 1; // must be 1
  bytes encrypted_metadata = 2; // AEAD(metadata_key, header_nonce, AD_metadata)
  bytes encrypted_payload = 3;  // AEAD(message_key, payload_nonce, AD_payload)
  bytes header_nonce = 4;       // 12 bytes
  uint64 ratchet_epoch = 5;     // monotonic, must not decrease
  optional bytes dh_public_key = 6;    // 32 bytes, present only on ratchet rotation
  optional bytes kyber_ciphertext = 7; // 1088 bytes, required if dh_public_key is set
  google.protobuf.Timestamp sent_at = 8;
}

message EnvelopeMetadata {
  uint64 message_index = 1;
  bytes payload_nonce = 2; // 12 bytes
  EnvelopeType envelope_type = 3;
  uint32 envelope_id = 4;
  optional string correlation_id = 5;
}

enum EnvelopeType {
  REQUEST = 0;
  RESPONSE = 1;
  NOTIFICATION = 2;
  HEARTBEAT = 3;
  ERROR_RESPONSE = 4;
}
